util/topsql/reporter/reporter.go

```diff
@@ -585,7 +653,12 @@ func (tsr *RemoteTopSQLReporter) doReport(data *ReportData) {
 		}
 	})
 	deadline := time.Now().Add(timeout)
-	if err := tsr.dataSink.TrySend(data, deadline); err != nil {
-		logutil.BgLogger().Warn("[top-sql] failed to send data to datasink", zap.Error(err))
+
+	tsr.dataSinks.mu.Lock()
+	defer tsr.dataSinks.mu.Unlock()
+	for ds := range tsr.dataSinks.m {
```

Comment:
You can copy out the dataSink list using a lock, then iterate each one without a lock. This can avoid deadlocks if `TrySend` is not carefully implemented.